cuda/preconditioner/jacobi_simple_apply_kernel.cu

```diff
@@ -0,0 +1,230 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2019, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/preconditioner/jacobi_kernels.hpp"
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+
+
+#include "core/base/extended_float.hpp"
+#include "core/matrix/dense_kernels.hpp"
+#include "core/preconditioner/jacobi_utils.hpp"
+#include "core/synthesizer/implementation_selection.hpp"
+#include "cuda/base/math.hpp"
+#include "cuda/base/types.hpp"
+#include "cuda/components/cooperative_groups.cuh"
+#include "cuda/components/diagonal_block_manipulation.cuh"
+#include "cuda/components/thread_ids.cuh"
+#include "cuda/components/uninitialized_array.hpp"
+#include "cuda/components/warp_blas.cuh"
+
+
+namespace gko {
+namespace kernels {
+namespace cuda {
+
+
+/**
+ * A compile-time list of block sizes for which dedicated apply kernels should
+ * be compiled.
+ */
+using compiled_kernels =
+    syn::value_list<int, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
```

Comment:
Is there any benefit in wrapping this as well in the `GINKGO_JACOBI_FULL_OPTIMIZATIONS` ? and enabling only a couple most used ones by default ?

Reply:
Yes there is, if I also limit these sizes to compile for the powers of two and 13 as a weird size for testing (1, 2, 4, 8, 13, 16, 32) here are the numbers:
+ `advanced_apply` : `Elapsed time: 19 s. (time), 0.000626 s. (clock)`
+ `generate` : `Elapsed time: 49 s. (time), 0.000495 s. (clock)`
+ other kernels: `Elapsed time: 5 s. (time), 0.000602 s. (clock)`
+ `simple_apply` : `Elapsed time: 15 s. (time), 0.000589 s. (clock)`

The main question is how much performance this also removes.

Reply:
I guess we probably have to benchmark that because I have no Idea how much performance it would cost us.
I was actually not even sure if it would work, but the test for `is_eligible` is actually a `<=`, so it would work to remove some numbers in between (if I remember correctly, @gflegar actually said that we need at least all powers of `2` to work properly).

Reply:
I added this together with the other improvements suggested.

Reply:
I thought a bit about it, and it makes sense to put the limited kernel number in there by default. If someone notices that the performance is less than expected, the person should find the parameter `GINKGO_JACOBI_FULL_OPTIMIZATIONS`, after which the maximum performance is achieved.
Therefore, I don't think benchmarking is that necessary any more.

Reply:
Well, I wanted the benchmark to serve more as a reference. Also, assuming the performance is terrible with the current options, we could try to find a better compromise between compile time and runtime for the default mode without optimizations.